### DISCUSSION TITLE
[BUG]: uploading files with multipart/form-data doesn't seem to work

### DISCUSSION THREAD

**Discussion by yomimono** [etsy/open-api#743] [Q&A]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When trying to upload a new file to a listing, I always get a 400 error with the body `{"error":"An error ocurred while uploading your file, please try again."}`  The listing is a digital listing and already has one file, uploaded via the web interface.

Curiously, my token doesn't seem to get checked, as I get the same error whether using a currently valid token or an expired one, so something's going wrong before that but I can't tell what. :thinking: 

Here's an example of the headers and body I've sent, with the token redacted. I'm using a very small file here for debugging, but I get the same result with "real" files (parseable `png`s and `pdf`s with the expected extensions).

I'm able to `GET https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{listing_id}/files` with the same token I'm trying below. The requested scopes include both `listings_r` and `listings_w`.

Here's what I POST to `https://api.etsy.com/v3/application/shops/{shop_id}/listings/{listing_id}/files`:

```
POST /v3/application/shops/redacted-shop/listings/redacted-listing/files HTTP/1.1
x-api-key: redacted-key
Authorization: Bearer redacted-id.redacted-token
Content-Type: multipart/form-data; boundary="9965-cXMSUA="


--9965-cXMSUA=
Content-Disposition: form-data; name="name"; 

dune
--9965-cXMSUA=
Content-Disposition: form-data; name="rank"; 

1
--9965-cXMSUA=
Content-Transfer-Encoding: binary
Content-Disposition: form-data; filename=file.txt; name="file"; 

cheese
--9965-cXMSUA=--


```

### Steps to reproduce

Post multipart form-data to /v3/application/shops/valid-shop-id/listings/valid-listing/files .

### Expected behavior

I would expect a `201` response with the metadata for the uploaded file, or an error message explaining what's wrong (as I received for many other error conditions, like not including a filename).

### Additional context

Answers on the HTTP level would be appreciated.

**Comment by yomimono** [comment]
Please let me know what additional information would help in diagnosing this issue. "works for me" or "example is wrong because $reason" would also be helpful, along with "I tried this on $date with $library in $language and it worked fine".

**Comment by yomimono** [comment]
In case this is useful, I'm able to successfully upload an image to the `/images` endpoint for the same listing using the same software, edited to push an `image` instead of a `file` to the respective endpoint, but still can't upload a new file.

**Comment by yomimono** [comment]
I've submitted a support request to Etsy asking whether I can expect an update on this issue within 7 days. Uploading files is a required feature (indeed, it's the core feature of the application I'm developing), and as this is a new application I certainly don't want to rewrite it for the previous API version.  If this can't be resolved I will need to move to another platform.

**Comment by jmarcv** [comment]
I know the docs say to use multiform, but the example in the Listings Tutorial uses:
Content-Type: application/x-www-form-urlencode
Maybe try that to see if the docs are wrong?

**Comment by baahar** [comment]
Was this issue resolved or did you find a way to solve this? 
I can upload an image using multiform without problems, but when trying to upload a file I'm getting the same error message. 

>

**Comment by baahar** [comment]
Ok, adding "Content-Type: data" before sending the byte array solved the problem for me.

**Comment by yomimono** [comment]
Thank you, I'll try that! I gave up on this API and haven't checked whether the issue is resolved, but your comment is encouraging.

**Comment by DeividRobim** [comment]
Did you find a solution? I keep getting {'error': 'A valid name must be provided with a new file.'}

**Reply by DeividRobim** [reply]
ETSY_API_BASEURL = "https://openapi.etsy.com/v3/application"
HEADERS = {
    "x-api-key": config["KEYSTRING"]
}

def upload_listing_file(data, file_path):
try:
shop_id = get_shop_id()

listing_id = data["listing_id"]

files = {"file": open(file_path, "rb")}
payload = {"name": "test", "rank": 2}

response = oauth_session.post(
f"{ETSY_API_BASEURL}/shops/{shop_id}/listings/{listing_id}/files",
headers=HEADERS,
files=files,
json=payload,
).json()

return response

except:
raise e

**Reply by jmarcv** [reply]
Deivid .... see abone:
Content-Type: application/x-www-form-urlencode

**Comment by heaven7** [comment]
Is there any working solution on this?
  
```python
    headers = {
        "Accept": "application/json",
        # "Content-Type": "application/x-www-form-urlencoded",
        "x-api-key": keystring,
    }
    file_contents = open(file_path, mode="rb")
    files = {'file': file_contents}
    params = {
        'name': 'test123'
    }
    refresh_url = "https://api.etsy.com/v3/public/oauth/token"
    client = OAuth2Session(keystring, token=token, auto_refresh_url=refresh_url, token_updater=token_saver)
    url = f"https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{listing_id}/files"
    r = client.post(url, files=files, json=params, headers=headers)
    
``` 
Here still the `A valid name must be provided with a new file.` error occurs.

Where to put the `name` right?

**Reply by Gareth064** [reply]
Try putting the file extension onto the end of the file name param. So `test123.jpg` and see how that works

**Reply by heaven7** [reply]
This is now working for me:

```python
def upload_listing_file(shop_id, listing_id, file_path):
    headers = {
        "Accept": "application/json",
        "x-api-key": keystring,
    }
    file_contents = open(file_path, mode="rb")
    file_name = file_path.split('/')[-1] # test123.jpg
    files = {'file': [file_name, file_contents, 'multipart/form-data']}
    data = {
        'name': file_name # test123.jpg
    }
    client = OAuth2Session(keystring, token=token, auto_refresh_url=refresh_url, token_updater=token_saver)
    url = f"https://openapi.etsy.com/v3/application/shops/{shop_id}/listings/{listing_id}/files"
    r = client.post(url, data=data, files=files, headers=headers)
    result = json.loads(r.text)
    return result
```

**Reply by PdubsCryptoWorld** [reply]
I'd like to say that I can confirm the upload listing code function when tailored to my specifics worked for me. Thank you very much heaven7
